Pod/Classes/NJHNavigationBarSelectorPageViewController.m

```diff
@@ -0,0 +1,228 @@
+
+/**
+ *  The ratio of the background selection view's height to its super view height
+ */
+static double const kNJHBackgroundSelectionViewVerticalRatio = 0.7;
+
+/**
+ *  The ratio of the background selection view's width to its super view width
+ */
+static double const kNJHBackgroundSelectionViewHorizontalRatio = 0.65;
+
+@interface NJHNavigationBarSelectorPageViewController ()
+@property (nonatomic) UINavigationItem *targetNavigationItem;
+@property (nonatomic) NSMutableArray *viewControllerArray;
+@property (nonatomic, getter=isPageScrolling) BOOL pageScrolling;
+@property (nonatomic, getter=hasInitialized) BOOL initialized;
+@property (nonatomic, readwrite) NSInteger currentPageIndex;
+@end
+
+@implementation NJHNavigationBarSelectorPageViewController
+
+- (instancetype)initWithPageViewControllers:(NSArray *)pageViewControllers navigationItem:(UINavigationItem *)navigationItem {
+    if (self = [super initWithTransitionStyle:UIPageViewControllerTransitionStyleScroll navigationOrientation:UIPageViewControllerNavigationOrientationHorizontal options:nil]) {
+        if (!navigationItem) {
+            self.targetNavigationItem = self.navigationItem;
+        } else {
+            self.targetNavigationItem = navigationItem;
+        }
+        
+        [self.viewControllerArray addObjectsFromArray:pageViewControllers];
+        self.delegate = self;
+        self.dataSource = self;
+    }
+    
+    return self;
+}
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    
+    self.navigationView.bounds = CGRectMake(0, 0, CGRectGetWidth(self.navigationController.navigationBar.frame) * kNJHBackgroundSelectionViewHorizontalRatio, CGRectGetHeight(self.navigationController.navigationBar.frame) * kNJHBackgroundSelectionViewVerticalRatio);
+    self.navigationView.center = CGPointMake(CGRectGetWidth(self.navigationController.navigationBar.frame) / 2, CGRectGetHeight(self.navigationController.navigationBar.frame) / 2);
+    
+    self.navigationView.labelTitles = [self generateButtonNamesFromViewControllers];
+    
+    self.targetNavigationItem.titleView = self.navigationView;
+}
+
+- (void)viewWillAppear:(BOOL)animated {
+    [super viewWillAppear:animated];
+    
+    // The self.viewControllers array is not constructed until viewWillAppear, so we have to do some initialization here
+    if (!self.hasInitialized) {
+        
+        [self setViewControllers:@[[self.viewControllerArray objectAtIndex:0]] direction:UIPageViewControllerNavigationDirectionForward animated:YES completion:nil];
+        
+        // http://stackoverflow.com/a/28242857/4080860
+        // The UIPageViewController does not expose the UIScrollView it uses to scroll between the view controllers it handles
+        // so we have to get it ourselves
+        
+        for (UIScrollView *view in self.view.subviews) {
+            if ([view isKindOfClass:[UIScrollView class]]) {
+                self.pageScrollView = view;
+                self.pageScrollView.delegate = self;
```

Comment:
I am guilty at doing the exact same thing (and worse). You have not found a better way?